Backup: add BackupRealtimeMessage component for displaying backup real-time info

Initsogar opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Resolves https://github.com/Automattic/jetpack-backup-team/issues/615

## Proposed Changes

* Add `BackupRealtimeMessage` component for displaying backup real-time info using the new fields introduced in https://github.com/Automattic/wp-calypso/pull/93885.
* Add `BackupRealtimeMessage` unit tests
* Render `BackupRealtimeMessage` on latest backup section in the backup page
* Conditionally render `BackupRealtimeMessage` under the `jetpack/backup-realtime-message`

![CleanShot 2024-08-26 at 20 17 26@2x](https://github.com/user-attachments/assets/1b8127a9-44bf-4989-9cd8-70f8d09b1fad)

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

For making clear daily vs real-time backups to end users.
Project thread: peaFOp-2Iy-p2

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
> [!NOTE]
> It could be tricky to test all scenarios on the UI because you might need to have a site with multiple backups and real-time events, so you can rely on unit tests for this PR.
> You can run unit tests locally by running:
> 
> `yarn run test-client client/components/jetpack/daily-backup-status/test/backup-realtime-message.jsx`

* Spin up a Jetpack Cloud live branch
* Enable the feature flag appending `flags=jetpack/backup-realtime-message` to the URL
* Pick a site with Jetpack VaultPress Backup with some backups and real-time events
* Navigate to the Backup page
* Search for a date where your last backup is a real-time event. If you can't find any, you could try doing some actions on your site like publishing/updating a post or uploading images.
* If your last backup is a real-time event, you should see a message like this:
  ![CleanShot 2024-08-27 at 15 47 34@2x](https://github.com/user-attachments/assets/7892bb6e-5bfb-41ea-b02a-43f368758462)
* If it is not a real-time event, you should not see the message:
  ![CleanShot 2024-08-27 at 15 49 33@2x](https://github.com/user-attachments/assets/f7d1a427-8346-442d-81dc-ee6dd84a66c1)




## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Initsogar committed:
feat: Add BackupRealtimeMessage component for displaying backup realtime information

Initsogar committed:
Render `BackupRealtimeMessage` on successful backup view

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into update/backup-realtime-changes

Initsogar committed:
Add unit tests for BackupRealtimeMessage component

mavegaf commented on client/components/jetpack/daily-backup-status/status-card/backup-realtime-message.tsx:
```diff
@@ -0,0 +1,83 @@
+	const translate = useTranslate();
+	const moment = useLocalizedMoment();
+	const backupDateMoment = moment( backupDate );
+	const today = moment();
```
Instead of comparing it with `today` we should compare it with the date of the backup selected

mavegaf commented on client/components/jetpack/daily-backup-status/status-card/backup-realtime-message.tsx:
```diff
@@ -0,0 +1,83 @@
+};
+
+export const BackupRealtimeMessage: FunctionComponent< Props > = ( {
+	backupDate,
```
We need as a parameter the date of the backup selected and the date of the base backup

mavegaf commented:
As an example, I selected August 16, where there is a full backup for that day, and the last backup is a real-time backup
<img width="714" alt="Screenshot 2024-08-27 at 20 24 38" src="https://github.com/user-attachments/assets/925d96c3-a19d-45f9-b93d-91aaa1345063">

In this screenshot, the base backup for that real-time is for the same day (`2024-08-16 11:44 AM`), but it says `11-day old full backup` because it is compared with today (August 27)

Initsogar commented:
> In this screenshot, the base backup for that real-time is for the same day (`2024-08-16 11:44 AM`), but it says `11-day old full backup` because it is compared with today (August 27)

@mavegaf Thanks for clarifying!
It means that the `we are using the full backup from today/yesterday` will never apply, or will not be very consistent.

I will suggest rephrasing them like:

- **Scenario 1 - The base backup is from the same date as the selected backup**
  - Before: `We are using the full backup from today(...)`
  - After: `We are using a full backup from this day(...)`
- **Scenario 2 - The base backup is from the day before as the selected backup**
  - Before: `We are using the full backup from yesterday(...)`
  - After: `We are using a full backup from the previous day(...)`
- **Scenario 3 - The base backup is from 2 or more days before as the selected backup**
  - (it remains the same): `We are using a #-day old full backup(...)`

Note that I'm also replacing `the` with `a`.

Any thoughts?

mavegaf commented:
> > In this screenshot, the base backup for that real-time is for the same day (`2024-08-16 11:44 AM`), but it says `11-day old full backup` because it is compared with today (August 27)
> 
> @mavegaf Thanks for clarifying! It means that the `we are using the full backup from today/yesterday` will never apply, or will not be very consistent.
> 
> I will suggest rephrasing them like:
> 
> * **Scenario 1 - The base backup is from the same date as the selected backup**
>   
>   * Before: `We are using the full backup from today(...)`
>   * After: `We are using a full backup from this day(...)`
> * **Scenario 2 - The base backup is from the day before as the selected backup**
>   
>   * Before: `We are using the full backup from yesterday(...)`
>   * After: `We are using a full backup from the previous day(...)`
> * **Scenario 3 - The base backup is from 2 or more days before as the selected backup**
>   
>   * (it remains the same): `We are using a #-day old full backup(...)`
> 
> Note that I'm also replacing `the` with `a`.
> 
> Any thoughts?

Yes, `today/yesterday` was a message only when the backup selected was the last one. I like the proposal, it's clear and simpler to implement 👍

Initsogar committed:
feat: Update BackupRealtimeMessage component to use baseBackupDate instead of backupDate

The BackupRealtimeMessage component has been updated to use the baseBackupDate prop instead of the backupDate prop. This change ensures that the correct backup date is displayed in the message. Also we are adding `selectedBackupDate` to use instead of `today` date.

Initsogar committed:
Refactor BackupSuccessful component to pass selectedBackupDate to BackupRealtimeMessage

Initsogar committed:
Remove redundant message scenario

Initsogar committed:
Replace `today` with `selectedBackupDate`

Initsogar committed:
Refactor BackupRealtimeMessage component to receive baseBackupDate as a Moment object

Initsogar committed:
Refactor BackupRealtimeMessage unit tests to consider latest changes

Initsogar committed:
Update test descriptions

Initsogar committed:
Improve unit test description grammar

Initsogar committed:
Improve showRealTimeMessage validation

Initsogar committed:
Remove `baseBackupDate` comment

Initsogar committed:
Refactor `BackupRealtimeMessage` conditions to improve readability and clarity

Initsogar commented:
@mavegaf Great, thanks for your feedback! I just made the changes based on that.
This is ready for a re-review :)

Also added the instructions to run the unit tests on the PR description:

`yarn run test-client client/components/jetpack/daily-backup-status/test/backup-realtime-message.jsx`

mavegaf commented:
@Initsogar With this PR I found that other cases need to be handled in the endpoint, so I've left it in Blocked/Hold until the changes in the API are ready

Initsogar committed:
Merge remote-tracking branch 'origin/trunk' into update/backup-realtime-changes

Initsogar committed:
Render BackupRealTimeMessage under a feature flag

Initsogar commented:
Since we are still polishing the API to fetch the right `base_rewind_id` and `rewind_step_count`, I just added a feature flag (`jetpack/backup-realtime-message`) for rendering the realtime message, so we could safely merge this PR and continue working on https://github.com/Automattic/wp-calypso/pull/94218 which has this one as dependency.

@mavegaf could you take another look?

Initsogar committed:
Move the feature flag outside the component

Initsogar committed:
Fix conditional rendering

mavegaf reviewed (approved):
👍

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16647738

Thank you @Initsogar for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16647738).
